src/hammer-tech/stackup.py

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+#  stackup.py
+#  Data structures to represent technology stackup options
+#
+#  See LICENSE for licence details.
+
+from enum import Enum
+from typing import Any, Callable, Iterable, List, NamedTuple, Optional, Tuple, Dict
+from hammer_utils import reverse_dict
+
+class RoutingDirection(Enum):
+    Vertical = 1
+    Horizontal = 2
+    Redistribution = 3
+
+    @classmethod
+    def __mapping(cls) -> Dict[str, "RoutingDirection"]:
+        return {
+            "vertical": RoutingDirection.Vertical,
+            "horizontal": RoutingDirection.Horizontal,
+            "redistribution": RoutingDirection.Redistribution
+        }
+
+    @staticmethod
+    def from_str(input_str: str) -> "RoutingDirection":
+        try:
+            return RoutingDirection.__mapping()[input_str]
+        except KeyError:
+            raise ValueError("Invalid routing direction: " + str(input_str))
+
+    def __str__(self) -> str:
+        return reverse_dict(RoutingDirection.__mapping())[self]
+
+    def opposite(self) -> "RoutingDirection":
+        if self == RoutingDirection.Vertical:
+            return RoutingDirection.Horizontal
+        elif self == RoutingDirection.Horizontal:
+            return RoutingDirection.Vertical
+        else:
+            return self
+
+# Note to maintainers: This is mirrored in schema.json- don't change one without the other
+
+# A tuple of wire width limit and spacing for generating a piecewise linear rule for spacing based on wire width
+# width_at_least: Any wires larger than this must obey the minSpacing rule
+# min_spacing: The minimum spacing for this bin. If a wire is wider than multiple entries, the worst-case (larger) minSpacing wins.
+class WidthSpacingTuple(NamedTuple('WidthSpacingTuple', [
+    ('width_at_least', float),
+    ('min_spacing', float)
+])):
+    __slots__ = ()
+
+    @staticmethod
+    def from_setting(d: dict) -> "WidthSpacingTuple":
+        return WidthSpacingTuple(
+            width_at_least=float(d["width_at_least"]),
+            min_spacing=float(d["min_spacing"])
+        )
+
+# A metal layer and some basic info about it
+# name: M1, M2, etc.
+# index: The order in the stackup (lower # is closer to the wafer)
```

Comment:
Maybe closer to the substrate? To me the whole thing is the wafer so all the metals are inside the wafer.

Reply:
lol yes, that's a horrible comment. I'll fix.